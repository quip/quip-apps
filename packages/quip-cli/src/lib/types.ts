interface Migration {
    version_number: number;
    js_file: string;
}

export interface Manifest {
    name: string;
    manifest_version: number;
    version_name: string;
    version_number: number;
    description?: string;
    toolbar_color?: string;
    diable_app_level_comments?: boolean;
    sizing_mode?: string;
    initial_width?: number;
    initial_height?: number;
    migrations?: Migration[];
}
