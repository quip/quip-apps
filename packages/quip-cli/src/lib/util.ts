import chalk from "chalk";
import {spawn} from "child_process";
import fs from "fs";
import minimatch from "minimatch";
import ncp from "ncp";
import path from "path";
import {println} from "./print";

export const runCmd = (cwd: string, command: string, ...args: string[]) => {
    return runCmdPromise(cwd, command, ...args)
        .then((stdout) => console.log(stdout))
        .catch((error) => {
            println(chalk`{red Command failed: ${command} ${args.join(" ")}}`);
            println(chalk`{red CWD: ${cwd}}`);
            println(chalk`{red ${error.stack}}`);
            process.exit(1);
        });
};

export const runCmdPromise = (
    cwd: string,
    command: string,
    ...args: string[]
): Promise<string> => {
    return new Promise((resolve, reject) => {
        const cmd = spawn(command, [...args], {
            cwd,
            stdio: ["inherit", "pipe", "inherit"],
        });
        let stdout = "";
        cmd.stdout.on("data", (d) => {
            stdout += d;
        });
        cmd.on("error", (error) => {
            reject(error);
        });
        cmd.on("close", () => {
            resolve(stdout);
        });
    });
};

export const readRecursive = async (
    dir: string,
    skip: string
): Promise<string[]> => {
    return new Promise((resolve, reject) => {
        fs.readdir(dir, (err, children) => {
            if (err) {
                return reject(err);
            }
            Promise.all<string[]>(
                children.map((child) => {
                    if (child) {
                        if (minimatch(child, skip)) {
                            return Promise.resolve([]);
                        }
                    }
                    return new Promise<string[]>((res, rej) => {
                        fs.stat(path.join(dir, child), (err, info) => {
                            if (err) {
                                return rej(err);
                            }
                            if (info.isDirectory()) {
                                res(
                                    readRecursive(
                                        path.join(dir, child),
                                        skip
                                    ).then((children) =>
                                        children.map((file) =>
                                            path.join(child, file)
                                        )
                                    )
                                );
                            } else {
                                res([child]);
                            }
                        });
                    });
                })
            ).then((childLists) => {
                resolve(
                    childLists.reduce((all, child) => [...all, ...child], [])
                );
            });
        });
    });
};

export const pathExists = (filePath: string): Promise<boolean> => {
    return new Promise((resolve, reject) => {
        fs.stat(filePath, (err) => {
            if (err && err.code === "ENOENT") {
                return resolve(false);
            } else if (err) {
                return reject(err);
            }
            resolve(true);
        });
    });
};

export const copy = (
    source: string,
    dest: string,
    options: ncp.Options = {}
): Promise<void> => {
    return new Promise((resolve, reject) =>
        ncp(source, dest, options, (err) => {
            if (err) {
                return reject(err);
            }
            resolve();
        })
    );
};
