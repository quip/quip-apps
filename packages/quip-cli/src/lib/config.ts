import fs from "fs";
import os from "os";
import path from "path";
import { pathExists } from "./util";

export const DEFAULT_SITE = "quip.com";
export const SKIP_SSL_FOR_SITES = new Set(["quip.codes"]);
export const NPM_BINARY_NAME = /^win/.test(process.platform)
    ? "npm.cmd"
    : "npm";

interface QLAConfigSite {
    accessToken: string;
}
export interface QLAConfig {
    _exists: boolean;
    sites: {
        [hostname: string]: QLAConfigSite;
    };
}

export const defaultConfigPath = () => path.join(os.homedir(), ".quiprc");

export const writeSiteConfig = async (
    configPath: string,
    site: string,
    config: QLAConfigSite
) => {
    const qlaConfig = await readConfig(configPath);
    qlaConfig.sites[site] = config;
    writeConfig(configPath, qlaConfig);
};

const writeConfig = (configPath: string, config: QLAConfig) => {
    let newConfig: Omit<QLAConfig, "_exists"> & { _exists?: boolean } = {
        ...config,
    };
    delete newConfig._exists;
    return fs.promises.writeFile(
        configPath,
        JSON.stringify(newConfig, null, 2),
        "utf-8"
    );
};

export const readConfig = async (configPath: string): Promise<QLAConfig> => {
    if (!(await pathExists(configPath))) {
        return { _exists: false, sites: {} };
    }
    const configStr = (await fs.promises.readFile(
        configPath,
        "utf-8"
    )) as string;
    try {
        const config = Object.assign({ _exists: true }, JSON.parse(configStr));
        config.sites = config.sites || {};
        return config;
    } catch (e) {
        throw new Error(`Corrupted config: ${e.message}`);
    }
};
