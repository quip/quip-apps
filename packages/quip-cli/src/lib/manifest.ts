import {flags} from "@oclif/command";
import fs from "fs";
import path from "path";
import prettier from "prettier";
import {Manifest} from "./types";
import {pathExists} from "./util";

export const manifestFlag = flags.string({
    char: "m",
    description:
        "A manifest.json file to add the migration to. By default, we'll use the first one we find.",
});

export const findManifest = async (
    dir: string
): Promise<string | undefined> => {
    const files = await fs.promises.readdir(dir);
    const dirs: string[] = [];
    for (const file of files) {
        if (file === "manifest.json") {
            return path.join(dir, file);
        }
    }
    for (const dir of dirs) {
        const manifest = await findManifest(dir);
        if (manifest) {
            return manifest;
        }
    }
};

interface FormattingInfo {
    indent: number;
}

const FIND_INDENT = /^{\n(\s+)[^\s]/;
const getManifestContent = async (
    manifestPath: string
): Promise<[Manifest, FormattingInfo]> => {
    if (!(await pathExists(manifestPath))) {
        throw new Error(`Manifest ${manifestPath} does not exist.`);
    }
    const content = (await fs.promises.readFile(
        manifestPath,
        "utf-8"
    )) as string;
    const formattingInfo = {indent: 2};
    const indentMatch = content.match(FIND_INDENT);
    if (indentMatch) {
        formattingInfo.indent = indentMatch[1].length;
    }
    try {
        return [JSON.parse(content), formattingInfo];
    } catch (e) {
        throw new Error(`Failed parsing manifest: ${manifestPath}`);
    }
};

export const getManifest = async (manifestPath: string): Promise<Manifest> => {
    const [manifest] = await getManifestContent(manifestPath);
    return manifest;
};

export const writeManifest = async (
    manifestPath: string,
    data: Partial<Manifest>
): Promise<Manifest> => {
    if (!(await pathExists(manifestPath))) {
        throw new Error(`Manifest ${manifestPath} does not exist.`);
    }
    const [manifest, {indent}] = await getManifestContent(manifestPath);
    // Note: should be obvious, but this won't do anything special for arrays.
    Object.assign(manifest, data);
    let prettierConfig = await prettier.resolveConfig(manifestPath);
    if (!prettierConfig) {
        prettierConfig = {tabWidth: indent};
    }
    await fs.promises.writeFile(
        manifestPath,
        prettier.format(JSON.stringify(manifest), {
            ...prettierConfig,
            parser: "json",
        }),
        "utf-8"
    );
    return manifest;
};
