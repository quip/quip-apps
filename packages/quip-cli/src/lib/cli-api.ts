import chalk from "chalk";
import FormData from "form-data";
import https from "https";
import fetch from "node-fetch";
import {readConfig, SKIP_SSL_FOR_SITES} from "../lib/config";
import {println} from "../lib/print";

interface ErrorResponse {
    error: string;
    response?: string;
}

interface AppVersionResponse {
    created_sequence: number;
    version_number: number;
    version_name: string;
    released: string;
}

export interface AppVersionsResponse {
    name: string;
    release: AppVersionResponse;
    development: AppVersionResponse;
    versions: AppVersionResponse[];
}

export interface AppListInfo {
    name: string;
    id: string;
}

export interface AppsListResponse {
    released: AppListInfo[];
    development: AppListInfo[];
    disabled: AppListInfo[];
}

const isError = (response: any): response is ErrorResponse =>
    response && response.error;

export const getStateString = () =>
    `${new Date().getTime()}${process.env.USER}`;

export const successOnly = async <T extends Object>(
    promise: Promise<T | ErrorResponse>,
    printJson: boolean
): Promise<T | false> => {
    let response: T | ErrorResponse;
    try {
        response = await promise;
    } catch (e) {
        response = {error: "Failed:", response: e.message};
    }
    if (printJson) {
        println(JSON.stringify(response));
    } else if (isError(response)) {
        println(chalk`
{red Error: ${response.error}}
{red ${response.response || ""}}`);
    } else {
        return response;
    }
    return false;
};

const cliAPI = async (configPath: string, site: string) => {
    const config = await readConfig(configPath);
    return async <T>(
        path: string,
        method?: "get" | "post",
        data?: {[key: string]: any} | FormData
    ): Promise<T | ErrorResponse> => {
        if (!config.sites[site]) {
            return {error: `Not logged in to ${site}`};
        }
        const {accessToken} = config.sites[site];
        return callAPI<T>(site, path, method, data, accessToken);
    };
};

const platformHost = (site: string) => {
    const subdomainCount = site.split(".").length - 2;
    return subdomainCount > 0 ? `platform-${site}` : `platform.${site}`;
};

export const callAPI = async <T = any>(
    site: string,
    path: string,
    method?: "get" | "post",
    data?: {[key: string]: any} | FormData,
    accessToken?: string
): Promise<T | ErrorResponse> => {
    let headers: {[name: string]: string} = {
        "Content-Type": "application/json",
    };
    if (accessToken) {
        headers["Authorization"] = `Bearer ${accessToken}`;
    }
    let body: string | FormData | undefined;
    if (data?.getHeaders) {
        headers = {...headers, ...data.getHeaders()};
        body = data as FormData;
    } else if (data) {
        body = JSON.stringify(data);
    }
    const request = await fetch(`https://${platformHost(site)}/cli/${path}`, {
        agent: new https.Agent({
            rejectUnauthorized: SKIP_SSL_FOR_SITES.has(site) ? false : true,
        }),
        method: method || "get",
        body,
        headers,
    });
    const raw = await request.text();
    try {
        return JSON.parse(raw);
    } catch (e) {
        return {error: "Invalid response", response: raw};
    }
};

export default cliAPI;
