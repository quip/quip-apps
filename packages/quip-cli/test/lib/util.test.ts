import path from "path";
import { readRecursive } from "../../src/lib/util";
describe("readRecursive", () => {
    test("basics", async () => {
        const files = await readRecursive(
            path.join(__dirname, "..", "fixtures", "readdir"),
            ""
        );
        expect(files).toMatchSnapshot();
    });
    test("skips", async () => {
        const files = await readRecursive(
            path.join(__dirname, "..", "fixtures", "readdir"),
            "test-subdir"
        );
        expect(files).toMatchSnapshot();
    });
});
