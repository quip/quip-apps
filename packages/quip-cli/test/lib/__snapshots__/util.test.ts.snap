// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readRecursive basics 1`] = `
Array [
  "test-dir/test-subdir/sub-sub-1",
  "test-dir/test-subdir/sub-sub-2",
  "test-dir/test-subfile",
  "test-file",
]
`;

exports[`readRecursive skips 1`] = `
Array [
  "test-dir/test-subfile",
  "test-file",
]
`;
